src/System/Taffybar/IconImages.hs

```diff
@@ -0,0 +1,88 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      : System.Taffybar.IconImages
+-- Copyright   : (c) Ivan A. Malison
+-- License     : BSD3-style (see LICENSE)
+--
+-- Maintainer  : Ivan A. Malison
+-- Stability   : unstable
+-- Portability : unportable
+-----------------------------------------------------------------------------
+
+module System.Taffybar.IconImages (
+  ColorRGBA,
+  scalePixbuf,
+  pixBufFromEWMHIcon,
+  pixelsARGBToBytesRGBA,
+  pixBufFromColor,
+  pixBufFromFile,
+  selectEWMHIcon
+) where
+
+import           Data.List
+import           Data.Ord
+import           Data.Word (Word8)
+import           Foreign.C.Types (CUChar(..))
+import           Foreign.Marshal.Array (newArray)
+import qualified Graphics.UI.Gtk as Gtk
+import           System.Information.EWMHDesktopInfo
+
+type ColorRGBA = (Word8, Word8, Word8, Word8)
+
+-- | Take the passed in pixbuf and ensure its scaled square.
+scalePixbuf :: Int -> Gtk.Pixbuf -> IO Gtk.Pixbuf
+scalePixbuf imgSize pixbuf = do
+  h <- Gtk.pixbufGetHeight pixbuf
+  w <- Gtk.pixbufGetWidth pixbuf
+  if h /= imgSize || w /= imgSize
+  then
+    Gtk.pixbufScaleSimple pixbuf imgSize imgSize Gtk.InterpBilinear
+  else
+    return pixbuf
+
+-- | Create a pixbuf from the pixel data in an EWMHIcon,
+-- scale it square, and set it in a GTK Image.
+pixBufFromEWMHIcon :: EWMHIcon -> IO Gtk.Pixbuf
+pixBufFromEWMHIcon EWMHIcon {width=w, height=h, pixelsARGB=px} = do
+  let pixelsPerRow = w
+      bytesPerPixel = 4
+      rowStride = pixelsPerRow * bytesPerPixel
+      sampleBits = 8
+      hasAlpha = True
+      colorspace = Gtk.ColorspaceRgb
+      bytesRGBA = pixelsARGBToBytesRGBA px
+  cPtr <- newArray $ map CUChar bytesRGBA
+  Gtk.pixbufNewFromData cPtr colorspace hasAlpha sampleBits w h rowStride
+
+-- | Create a pixbuf with the indicated RGBA color,
+-- scale it square, and set it in a GTK Image.
+pixBufFromColor :: Int -> ColorRGBA -> IO Gtk.Pixbuf
+pixBufFromColor imgSize (r,g,b,a) = do
+  let sampleBits = 8
+      hasAlpha = True
+      colorspace = Gtk.ColorspaceRgb
+  pixbuf <- Gtk.pixbufNew colorspace hasAlpha sampleBits imgSize imgSize
+  Gtk.pixbufFill pixbuf r g b a
+  return pixbuf
+
+-- | Convert a list of integer pixels to a bytestream with 4 channels.
+pixelsARGBToBytesRGBA :: [Int] -> [Word8]
```

Comment:
btw, if youre looking for places to improve performance further, this is a good place.
i wrote this more to show exactly how it worked than to be performant.